armory/utils/metrics.py

```diff
@@ -431,6 +431,192 @@ def _check_object_detection_input(y, y_pred):
         )
 
 
+def _intersection_over_union(box_1, box_2):
+    """
+    Assumes format of [y1, x1, y2, x2]
+    """
+    assert box_1[2] >= box_1[0]
+    assert box_2[2] >= box_2[0]
+    assert box_1[3] >= box_1[1]
+    assert box_2[3] >= box_2[1]
+
+    if sum([mean < 1 and mean >= 0 for mean in [box_1.mean(), box_2.mean()]]) == 1:
+        logger.warning(
+            "One set of boxes appears to be normalized while the other is not"
+        )
+
+    # Determine coordinates of intersection box
+    x_left = max(box_1[1], box_2[1])
+    x_right = min(box_1[3], box_2[3])
+    y_top = max(box_1[0], box_2[0])
+    y_bottom = min(box_1[2], box_2[2])
+
+    intersect_area = max(0, x_right - x_left) * max(0, y_bottom - y_top)
+    if intersect_area == 0:
+        return 0
+
+    box_1_area = (box_1[3] - box_1[1]) * (box_1[2] - box_1[0])
+    box_2_area = (box_2[3] - box_2[1]) * (box_2[2] - box_2[0])
+
+    iou = intersect_area / (box_1_area + box_2_area - intersect_area)
+    assert iou >= 0
+    assert iou <= 1
+    return iou
+
+
+def object_detection_mAP(list_of_ys, list_of_y_preds):
```

Comment:
I wonder if mAP is a confusing name since it returns AP per class rather than mAP?

Reply:
yea, I wavered on this. I was considering just returning the actual mAP scalar value, but I thought it'd be useful to have the AP's by class logged to the user by default. 

I held off on naming it something like `AP_per_class` only because mAP is a ubiquitous term for object detection metrics. Any opinion @davidslater ?

Reply:
I would rather name it specifically what it is. In the scenario configs we can show how you can generate mAP from that.

Reply:
Renaming to `object_detection_AP_per_class()`. Currently, only the mAP scalar value is saved to the results dict that's written to the output json. I'm modifying this so that both the `average_precision_by_class` dict and the mAP scalar value are saved (so you don't have to manually compute mean from the dict values). They will both also be logged (as is currently implemented).